### DISCUSSION TITLE
Better error handling

### DISCUSSION THREAD

**Discussion by florian-lefebvre** [withastro/roadmap#925] [Proposal]
### Body

# Summary

Document `500.astro` and allow retrieving the thrown error in it.

# Background & Motivation

Error handling is an important part of an SSR application. Things can go wrong and we don't want to show our users the scary default page of Vercel, Netlify etc.

The `500.astro` page had been requested in #610 and even though no answer has been given nor documentation added, this page is in fact supported. But it currently lacks the ability to retrieve the error and handle it properly (show some data, report etc).

# Goals

- Document the existence of `500.astro`
- Retrieve the thrown error
- Be able to use the `500.astro` page in dev to test it properly

# Example

We can take example on [Nuxt](https://nuxt.com/docs/getting-started/error-handling#error-page) and [Next.js](https://nextjs.org/docs/app/building-your-application/routing/error-handling).

The error could be provided by props:
```astro
---
// src/pages/500.astro
interface Props {
	error: unknown
}

const { error } = Astro.props
```

Using the `500.astro` instead of the dev overlay could be made possible using a flag, for example an environment variable or a new argument for `astro dev`

**Comment by florian-lefebvre** [comment]
POC at https://github.com/withastro/astro/pull/11134

**Comment by millette** [comment]
Were 401, 403, 4xx (400) pages discussed as well? I just tried to add a 403.astro page but it's not being picked up (with node adapter).

**Reply by florian-lefebvre** [reply]
Nope only 404 and 500 are. However I know there's an integration for it https://github.com/JohannesKlauss/astro-error-pages

**Reply by millette** [reply]
Merci for the pointer.

**Reply by rburgst** [reply]
the link produces a 404 now https://github.com/JohannesKlauss/astro-error-pages

**Reply by rburgst** [reply]
I would really appreciate having https://github.com/withastro/astro/blob/main/packages/astro/src/core/constants.ts#L63 also contain the usual error codes: 400, 401, 403! Working around with redirects to /400 is just not cutting it as it will give the caller the wrong initial result code.

**Reply by florian-lefebvre** [reply]
fyi you can implement this in userland by checking the response status and calling `next` with the url you want :+1:

**Reply by rburgst** [reply]
While this is a good first step, I would really think that extending astro to support other errors would increase symmetry. Currently its not really clear why 404 and 500 are any different from the others.
Also note that its a bit tricky to get the final response with the correct error code.

**Reply by florian-lefebvre** [reply]
I know we discussed allowing `3xx.astro` to style redirects, so yeah maybe it makes sense to have `4xx.astro` and `5xx.astro` with a `status` prop. I'll discuss it

**Reply by ematipico** [reply]
@rburgst mainly because those are the most known and used error codes in web dev. I don't have all the history but that's my best guess.

**Reply by florian-lefebvre** [reply]
We discussed it yesterday. We're not against but it would be better to have it as a distinct RFC since it's really different from my original proposal

**Reply by rburgst** [reply]
Can you create one, please? I am not sure what is needed to write a proper RFC

**Reply by ematipico** [reply]
@rburgst  you can create one yourself, just use this template: https://github.com/withastro/roadmap/blob/main/stage-1--discussion-template.md
